### PR TITLE
Editor: Set document title using DocumentHead component

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to assign the title of the post editor screen using the [`<DocumentHead />` component](https://github.com/Automattic/wp-calypso/tree/master/client/components/data/document-head). This will enable accurate titles to be shown when editing custom post types, e.g. "Edit Project" for portfolios instead of "Edit".

Before|After
---|---
Before|After

__Implementation notes:__

- ~~Removes a few conflicting placeholder Redux actions proposed to be removed already in #5438~~ (merged as part of #5438)
- Fixes issue where the editor post state was not correctly updated to reflect new post ID after save (in production, you can verify that saving a new post does not change the `<EditorPostType />` label from "New Post" to "Post" until after a complete refresh).
- Resets post edits via component unmounting instead of route exit, which is more consistent with the current implementation and prevents an issue where the title briefly changes from "Edit Post" to "Edit" when leaving the editor

__Testing instructions:__

Verify that the post editor title correctly reflects the current status of editing:
- "New Post" for editing a new post
- "Edit Post" for editing an existing post
- "New Page" for editing a new page
- "Edit Page" for editing an existing page
- "New" for editing a new custom post type if the post type labels have not loaded from the REST API
- "Edit" for editing an existing custom post type if the post type labels have not loaded from the REST API
- The custom post type object `labels.new_item` when editing a new custom post type
- The custom post type object `labels.edit_item` when editing an existing custom post type

Ensure that the issue fixed by #4539 in resetting edits upon leaving the editor remains fixed.

### CODE REVIEW

**Comment by apeatling** [comment]
@aduth Who is a good person to review this?

**Comment by aduth** [comment]
Rebased to resolve conflicts. @nylen would you mind taking a look over these changes?

File: client/post-editor/editor-document-head/index.jsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		<div>
+			{ siteId && 'page' !== type && 'post' !== type && (
+				<QueryPostTypes siteId={ siteId } />
+			) }
+			<DocumentHead title={ title } />
+		</div>
+	);
+}
+
+EditorDocumentHead.propTypes = {
```
Comment (timmyc) [inline client/post-editor/editor-document-head/index.jsx]: 🌟

**Comment by timmyc** [comment]
stats_ _wordpress_com

Looking good and tests out well.  Confirmed no regressions with #4539

**Pull request merged**